src/core/control/jobs/RenderJob.cpp

```diff
@@ -79,12 +80,32 @@ void RenderJob::run() {
 
         std::lock_guard lock(this->view->drawingMutex);
         std::swap(this->view->crBuffer, newBuffer);
+        repaintPage();
     } else {
-        for (Rectangle<double> const& rect: rerenderRects) { rerenderRectangle(rect); }
+        for (Rectangle<double> const& rect: rerenderRects) {
+            rerenderRectangle(rect);
+            repaintPageArea(rect.x, rect.y, rect.x + rect.width, rect.y + rect.height);
+        }
     }
+}
+
+void RenderJob::repaintPage() {
+    int x1 = view->getX();
+    int y1 = view->getY();
+    int x2 = x1 + view->getDisplayWidth();
+    int y2 = y1 + view->getDisplayHeight();
+    repaintWidgetArea(view->xournal->getWidget(), x1, y1, x2, y2);
+}
+
+void RenderJob::repaintPageArea(double x1, double y1, double x2, double y2) {
+    double zoom = view->xournal->getZoom();
+    int x = view->getX();
+    int y = view->getY();
+    repaintWidgetArea(view->xournal->getWidget(), x + std::floor(zoom * x1), y + std::floor(zoom * y1), x + std::ceil(zoom * x2), y + std::ceil(zoom * y2));
```

Comment:
The math doesn't look right to me here.
* `x + std::floor(zoom * x1)` here is the same as `view->getX() + std::floor(zoom * view->getX())` due to how the function is called. However, `view->getX()` is the top-left corner of the page in layout coordinates and does not need to be scaled by the zoom. Shouldn't the correct value be `view->getX()`?
* The x2 parameter is calculated as `x + std::ceil(zoom * x2)` which is equivalent to `view->getX() + std::ceil(zoom * view->getX() + view->getDisplayWidth())`. Shouldn't the correct value be `view->getX() + view->getDisplayWidth()`?

Reply:
Oh, sorry, I got confused. I misread the call in `repaintWidgetArea` in `repaintPage` as `repaintPageArea`. This looks fine to me on a second look.